modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+                           'Time'       => (new \DateTime($collection['time']))->format('Y-m-d H:i'),
+                           'Comments'   => $collection['comments'],
+                          );
+        $collectionData['Data'] = $collection['data'] ? json_encode($collection['data']) : null;
+
+        $preparation = $specimen->getPreparation();
+        if (isset($preparation)) {
+            $preparationData         = array(
+                                        'SpecimenID'         => $specimen['id'],
+                                        'SpecimenProtocolID' => $preparation['protocolId'],
+                                        'CenterID'           => $preparation['centerId'],
+                                        'Date'               => (new \DateTime($preparation['date']))->format('Y-m-d H:i'),
+                                        'Time'               => (new \DateTime($preparation['time']))->format('Y-m-d H:i'),
+                                        'Comments'           => $preparation['comments'],
+                                       );
+            $preparationData['Data'] = $preparation['data'] ? json_encode($preparation['data']) : null;
+        }
+
+        $analysis = $specimen->getAnalysis();
+        if (isset($analysis)) {
+            $analysisData         = array(
+                                     'SpecimenID'       => $specimen['id'],
+                                     'SpecimenMethodID' => $analysis['methodId'],
+                                     'CenterID'         => $analysis['centerId'],
+                                     'Date'             => (new \DateTime($analysis['date']))->format('Y-m-d H:i'),
+                                     'Time'             => (new \DateTime($analysis['time']))->format('Y-m-d H:i'),
+                                     'Comments'         => $analysis['comments'],
+                                    );
+            $analysisData['Data'] = $analysis['data'] ? json_encode($analysis['data']) : null;
+        }
+
+        return $data = array(
+                        'biobank_specimen'             => $specimenData,
+                        'biobank_specimen_freezethaw'  => $freezeThawData,
+                        'biobank_specimen_collection'  => $collectionData,
+                        'biobank_specimen_preparation' => $preparationData,
+                        'biobank_specimen_analysis'    => $analysisData,
+                       );
+    }
+
+    /**
+     * This function takes an array that resulted from an SQL query and
+     * instantiates it as a Specimen Object.
+     *
+     * @param array $data              Values to be converted to array.
+     * @param array $parentSpecimenIds IDs of the parent Specimen.
+     *
+     * @return array
+     */
+    private function getArrayFromSQL(array $data, array $parentSpecimenIds) : array
+    {
+        // TODO: check type before typecasting.
```

Comment:
They will always be `string` if it's coming from a MySQL query.

Reply:
true that